Capability Request available in global on zome function call

zippy opened the pull request:
## PR summary

Context: This is a pre-work PR getting ready for hdk::grant and hdk:claim.  It achieves:

- one public facing thing, which is a new Zome global  CAPABILITY_REQ which makes the request available during all zome calls.  This is important for doing validation of provenance for a zome call that wants to create a capability grant.
- one code reordering thing, which is moving the CapabilityRequest struct into core_types
- app-spec test of CAPABILITY_REQ

## changelog

Please check one of the following, relating to the [CHANGELOG](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG.md)

- [x] this is a code change that effects some consumer (e.g. zome developers) of holochain core so there is an 'Unreleased' CHANGELOG item, with the format `- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)`
- [ ] this is not a code change, or doesn't effect anyone outside holochain core development

Connoropolous commented on core_types/src/dna/capabilities.rs:
```diff
@@ -44,3 +25,47 @@ pub enum CapabilityType {
+//--------------------------------------------------------------------------------------------------
+
+/// a struct to hold the capability information needed to make any capability request,
+/// namely the provenance of the request (the agent address an signature) and the
```
```suggestion
/// namely the provenance of the request (the agent address and signature) and the
```

Connoropolous commented on CHANGELOG.md:
```diff
@@ -28,6 +28,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Adds CAPABILITY_REQ global into the zome context. This makes the request available during all zome calls and is important for doing validation of provenance for a zome call that wants to create a capability grant. [#1273](https://github.com/holochain/holochain-rust/pull/1273)
```
@zippy can you reword this?
This makes no sense to me...  I'll break down what I don't understand specifically if that helps...
What's "global into the zome context"?
"the request" which request?
"a capability grant" What's a capability grant, from the perspective of someone who's never heard of it before?

struktured commented:
It could considering reference prior work on capabilities- I believe we
mean capability in the sense of this:
https://en.wikipedia.org/wiki/Capability-based_security.

On Wed, Apr 17, 2019 at 11:45 AM Connor Turland <notifications@github.com>
wrote:

> *@Connoropolous* requested changes on this pull request.
> ------------------------------
>
> In CHANGELOG.md
> <https://github.com/holochain/holochain-rust/pull/1273#discussion_r276309823>
> :
>
> > @@ -28,6 +28,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
>
>  ### Added
>
> +- Adds CAPABILITY_REQ global into the zome context. This makes the request available during all zome calls and is important for doing validation of provenance for a zome call that wants to create a capability grant. [#1273](https://github.com/holochain/holochain-rust/pull/1273)
>
> @zippy <https://github.com/zippy> can you reword this?
> This makes no sense to me... I'll break down what I don't understand
> specifically if that helps...
> What's "global into the zome context"?
> "the request" which request?
> "a capability grant" What's a capability grant, from the perspective of
> someone who's never heard of it before?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/holochain/holochain-rust/pull/1273#pullrequestreview-227841066>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE01okhc6iM01YaIeV3DwppgcOqbF6ZRks5vh0GsgaJpZM4czKeD>
> .
>

zippy commented:
@Connoropolous I updated the changelog text and added a link to the architecture ADR to explain capabilities grant rather than trying to do it in the change-log text itself, as that would be too verbose I think.

AshantiMutinta commented on core/src/context.rs:
```diff
@@ -279,33 +283,31 @@ impl Context {
+                if grant.cap_type() == CapabilityType::Public
+                    && grant.id() == ReservedCapabilityId::Public.as_str()
+                {
+                    return Ok(addr);
```
Just an added comment but we could probably use a fold() to make it look more idiomatic without the early return but no qualms about the way it is done here

AshantiMutinta commented on core_types/src/entry/cap_entries.rs:
```diff
@@ -21,6 +21,32 @@ pub enum CapabilityType {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "hc_public" => Ok(ReservedCapabilityId::Public),
+            _ => Err("Cannot convert string to ReservedCapabilityId"),
```
I like this
